### PR TITLE
Multimethods for mathematical functions

### PR DESCRIPTION
**Pull request by joaosferreira**
This pull request further adds multimethods that target [mathematical functions](https://numpy.org/doc/stable/reference/routines.math.html). I'll name the multimethods added according to the NumPy docs sectioning:

**Trigonometric functions**

- `degrees`
- `radians`
- `unwrap`

**Rounding**

- `around` (and its aliase `round_`)
- `fix`

**Sums, products, differences**

- `cumprod`
- `cumsum`
- `nancumprod`
- `nancumsum`
- `ediff1d`
- `cross`
- `trapz`

**Other special functions**

- `i0`
- `sinc`

**Arithmetic operations**

- `float_power`

**Handling complex numbers**

- `angle`
- `real`
- `imag`
- `conjugate`

**Miscellaneous**

- `convolve`
- `clip`
- `nan_to_num`
- `real_if_close`
- `interp`

Some thoughts about things to address moving forward with the PR:

1. There are no default implementations in this initial commit. I intend to fix the failing tests first before I start pushing some. The failing tests are all in the XND backend and are almost all the same. 
2. I noticed that the multimethod for `diff` has a different signature from the NumPy's function (the latter has more arguments). I'm wondering if these were added after the unumpy's multimethod was written.
3. I think some older multimethods for ufuncs don't have tests so I might add those as well.

### CODE REVIEW

File: unumpy/tests/test_numpy.py [hunk 1]
```diff
@@ -347,9 +363,10 @@ def test_array_creation(backend, method, args, kwargs):
         (np.divide, ([6, 1], [3, 2]), {}, [2.0, 0.5]),
         (np.true_divide, ([6, 1], [3, 2]), {}, [2.0, 0.5]),
         (np.power, ([2, 3], [3, 2]), {}, [8, 9]),
+        (np.float_power, ([2, 3], [3, 2]), {}, [8, 9]),
         (np.positive, ([1, -2],), {}, [1, -2]),
         (np.negative, ([-2, 3],), {}, [2, -3]),
-        (np.conj, ([1.0 + 2.0j, -1.0 - 1j],), {}, [1.0 - 2.0j, -1.0 + 1j]),
+        (np.conjugate, ([1.0 + 2.0j, -1.0 - 1j],), {}, [1.0 - 2.0j, -1.0 + 1j]),
```
Comment (peterbell10) [inline unumpy/tests/test_numpy.py:417]: You should test the alias as well.

File: unumpy/_multimethods.py [hunk 2]
```diff
@@ -1379,7 +1389,20 @@ def _ediff1d_argreplacer(args, kwargs, dispatchables):
...
 
 
-@create_numpy(_ediff1d_argreplacer)
+def _ediff1d_default(ary, to_end=None, to_begin=None):
+    ary = ravel(ary)
+
+    diffs = ary[1:] - ary[:-1]
+
+    if to_end is None:
+        to_end = asarray([], dtype=ary.dtype)
```
Comment (hameerabbasi) [inline unumpy/_multimethods.py]: What's going on here? Concatenating an empty array does nothing.

Comment (joaosferreira) [reply unumpy/_multimethods.py]: You are right, I could do the other way around like:

```python
if to_end is not None:
    diffs = concatenate([diffs, ravel(to_end)])
```

Seems better this way!

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: Append to a list, check for trivial case and return early.

**Comment by joaosferreira** [comment]
There's a type error arising from the call to `where` in `nancumprod` and `nancumsum` defaults. Should I ignore them?

**Comment by hameerabbasi** [comment]
Yes, just add `type: ignore`

**Comment by joaosferreira** [comment]
The default for `interp` was based on NumPy's [implementation](https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/function_base.py#L1289-L1423) and uses the [equation of slopes](https://en.wikipedia.org/wiki/Linear_interpolation#Linear_interpolation_between_two_known_points) to compute the y-coordinates element-wise.

I'm still not sure how to add the `left` and `right` cases. Also, it needs to take into consideration complex numbers.

**Comment by peterbell10** [comment]
Instead of clipping `x`, couldn't you do something like this:

```python
# Where we return currently
result = (y0 * x1_dist + y1 * x0_dist) / x_dist 

left = fp[0] if left is None else left
right = fp[-1] if right is None else right
result = where(x < xp[0], left, result)
return where(x > xp[-1], right, result)
```

You will need to fixup `idxs` as well to make sure there are no out of bounds accesses when interpolating.

**Comment by joaosferreira** [comment]
I've pushed the changes to `_interp_default` that add the `left` and `right` cases. It also works for complex numbers.

**Comment by joaosferreira** [comment]
NumPy's implementation for [`unwrap`](https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/function_base.py#L1487-L1541) uses item assignment which is not supported in backends with read-only arrays. Maybe there's a way to implement the default without using item assignment?

**Comment by hameerabbasi** [comment]
I'd recommend you first do the calculation as if `slice1` was only used for reading, not assignment, then, when it's being used for assignment: https://github.com/numpy/numpy/blob/v1.19.0/numpy/lib/function_base.py#L1540

I'd recommend you replace that line with a concatenate in the appropriate axis with a sub-array in the appropriate axis. A 1D version is below, I'll leave it as an exercise to come up with N-D (hint: tuple with `slice(None)` and `0`, and axis in `concatenate`).

```
up_slice1 = p[slice1] + ph_correct.cumsum(axis)
up = concatenate([p[0], up_slice1])
```

File: unumpy/_multimethods.py [hunk 3]
```diff
@@ -1872,7 +1872,36 @@ def rot90(m, k=1, axes=(0, 1)):
...
 
 
-@create_numpy(_self_argreplacer)
+def _unwrap_default(p, discont=3.141592653589793, axis=-1):
+    nd = ndim(p)
+
+    dd = diff(p, axis=axis)
+
+    slice0 = [slice(None)] * nd
+    slice0[axis] = 0
```
Comment (peterbell10) [inline unumpy/_multimethods.py]: You can use `slice(0, 1)` to grab the 0 value without also collapsing the dimension. That will avoid the need to use `expand_dims` later.

Comment (joaosferreira) [reply unumpy/_multimethods.py]: Good idea, that way we avoid an unnecessary multimethod call.

File: unumpy/_multimethods.py [hunk 4]
```diff
@@ -1770,3 +1870,193 @@ def roll(a, shift, axis=None):
...
+@create_numpy(_self_out_argreplacer)
+@all_of_type(ndarray)
+def around(a, decimals=0, out=None):
+    return (a, mark_non_coercible(out))
+
+
+round_ = around
+
+
+def _fix_default(x, out=None):
```
Comment (peterbell10) [inline unumpy/_multimethods.py]: I think this is the same as [`trunc`](https://numpy.org/doc/stable/reference/generated/numpy.trunc.html).

>The truncated value of the scalar x is the nearest integer i which is closer to zero than x is.

**Comment by hameerabbasi** [comment]
We've merged the fix we talked about in the meeting in Quansight-Labs/uarray#246. Pushing an empty commit with `git commit --allow-empty` (or any changes) should fix CI if there are no other breaking changes.

File: unumpy/_multimethods.py (hunk 4 above)
Comment (joaosferreira) [reply unumpy/_multimethods.py]: They seem to be the same, yes. Refactor the default in terms of `trunc` then?  Something like:

```python
def _fix_default(x, out=None):
    x = trunc(x)

    if out is None:
        return x
    else:
        copyto(out, x)
```

Comment (hameerabbasi) [reply unumpy/_multimethods.py]: Why not just `return trunc(x, out=out)`?

Comment (joaosferreira) [reply unumpy/_multimethods.py]: You are right! Forgot about the `out` kwarg in ufuncs. 😅

**Comment by joaosferreira** [comment]
I think this can be merged now.

**Comment by hameerabbasi** [comment]
Thanks, @joaosferreira!

**Pull request merged**